src/config.ts

```diff
@@ -0,0 +1,186 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AuthenticationConfig, Common, ServiceConfig} from '../third_party/types/common-types';
+
+const common: Common = require('@google-cloud/common');
+const extend = require('extend');
+
+// Configuration for Profiler.
+export interface Config extends AuthenticationConfig {
+  // Log levels: 0-disabled,1-error,2-warn,3-info,4-debug.
+  // Defaults to value of 1.
+  logLevel?: number;
+
+  // Specifies the service with which profiles from this application will be
+  // associated.
+  serviceContext?: {
+    // Name of the service under which the profiled data will be recorded and
+    // exposed in the UI for the project.
+    // You can specify an arbitrary string, see deployment.target at
+    // https://github.com/googleapis/googleapis/blob/master/google/devtools/cloudprofiler/v2/profiler.proto
+    // for restrictions.
+    // The string should be the same across different replicas of your service
+    // so that a globally constant profiling rate is maintained.
+    // service defaults to the value in the environment variable GAE_SERVICE.
+    service?: string;
+
+    // Version of the service. It can be an arbitrary string. Stackdriver
+    // Profiler profiles each version of each service in each zone once per
+    // minute.
+    // Defaults to the value in the environment variable GAE_VERSION.
+    // If there is no value for GAE_VERSION, it defaults to an empty string.
+    version?: string;
+  };
+
+  // Virtual machine instance to associate profiles with instead of the one
+  // read from the VM metadata server.
+  // Defaults to an empty string.
+  instance?: string;
+
+  // Zone to associate profiles with instead of the one read from the VM
+  // metadata server.
+  // Defaults to an empty string.
+  zone?: string;
+
+  // When true, CPU profiling will be disabled.
+  // Defaults to false.
+  disableCpu?: boolean;
+
+  // When true, heap profiling will be disabled.
+  // Defaults to false.
+  disableHeap?: boolean;
+}
+
+// Default values for configuration for a profiler.
+const defaultConfig: Config = {
+  projectId: '{{projectId}}',
+  logLevel: 1,
+  serviceContext: {
+    service: undefined,
+    version: '',
+  },
+  disableHeap: false,
+  disableCpu: false,
+};
+
+// initConfig sets unset values in the configuration to the value retrieved from
+// environment variables or to the default value specified in defaultConfig.
+function initConfig(config: Config): Config {
+  config = common.util.normalizeArguments(null, config);
+
+  const envConfig = {
+    logLevel: process.env.GCLOUD_PROFILER_LOGLEVEL,
+    projectId: process.env.GCLOUD_PROJECT,
+    serviceContext: {
+      service: process.env.GAE_SERVICE,
+      version: process.env.GAE_VERSION,
+    }
+  };
+
+  config = extend(true, {}, defaultConfig, envConfig, config);
+  return config;
+}
+
+export class ConfigImpl implements Config {
```

Comment:
Does this class need to be exported? I assume clients don't need to use it, do they?

I think it would make sense to do something basically close to what Trace interface does:

* A config interface with documentation and all the fields as optional: https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/212f45fd2c1df0ac01c16c8599e7a1792478c719/config.ts#L21

* A const defaultConfig with default values and no docs: https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/212f45fd2c1df0ac01c16c8599e7a1792478c719/config.ts#L133

* initConfig method in index.ts which merges the configurations like https://github.com/GoogleCloudPlatform/cloud-trace-nodejs/blob/212f45fd2c1df0ac01c16c8599e7a1792478c719/index.ts#L78 which is 
```
  // Configuration order of precedence:
  // Default < Environment Variable Set Configuration File < Project
  var config = extend(true, {}, defaultConfig, envSetConfig,
    projectConfig, envConfig);
```

(except that I am not a fan projectConfig name but whatever - consistency is probably more important).